doc/user-manual/language/sort-system.lagda.rst

```diff
@@ -0,0 +1,193 @@
+--------
+
+``univSort`` returns the successor sort of a given sort.
+
+Example: the sort of the function type ``∀ {A} → A`` with normal form
+``{A : _5} → A`` evaluates to ``funSort (univSort _5) _5`` where:
+
+* ``univSort _5`` is the sort where the sort of ``A`` lives, ie. the
+  successor level of ``_5``.
+
+Note that ``univSort`` appled to ``Setω`` is well-defined only if the option
+``--omega-in-omega`` is enabled in the agda file.
+
+We list below all the possible computations that ``univSort`` can perform:
+
+.. code-block::
+
+  univSort (Set a)  = Set (lsuc a)
+  univSort (Prop a) = Set (lsuc a)
+  univSort Setω     = Setω                      (only if --omega-in-omega is enabled)
+
+.. _piSort:
+
+piSort
+------
+
+Similarly, ``piSort s1 s2`` is a constructor that computes the sort of
+a Π-type given the sort ``s1`` of its domain and the sort ``s2`` of its
+codomain as arguments.
+
+To understand how ``piSort`` works in general, we set the following scenario:
+
+* ``sA`` and ``sB`` are two (possibly different) sorts.
+* ``A : sA``, meaning that ``A`` is a type that has sort ``sA``.
+* ``x : A``, meaning that ``x`` has type ``A``.
+* ``B : sB``, meaning that ``B`` is a type (possibly different than ``A``) that
+  has sort ``sB``.
+
+Under these conditions, we can build the dependent function type
+``(x : A) → B : piSort sA (λ x → sB)``. This type signature means that the
+dependent function type ``(x : A) → B`` has a (possibly unknown) but
+well-defined sort ``piSort sA sB``, specified in terms of the element
+``x : A`` and the sorts of its domain and codomain.
+
+We list below all the possible computations that ``piSort`` can perform:
+
+.. code-block::
+
+  piSort s1 (λ x → s2) = funSort s1 s2          (if x does not occur freely in s2)
+  piSort s1 (λ x → s2) = Setω                   (if x occurs freely in s2)
+
+All these constructors do not represent new sorts but instead, they compute
+to the right sort once their arguments are known.
+
+More examples:
+
+* ``piSort Level (λ l → Set l)`` evaluates to ``Setω``
+* ``piSort (Set l) (λ _ → Set l')`` evaluates to ``Set (l ⊔ l')``
+* ``univSort (Set l)`` evaluates to ``Set (lsuc l)``
+* The function type ``∀ {A} → ∀ {B} → B → A → B`` with normal form
+  ``{A : _9} {B : _7} → B → A → B`` evaluates to
+  ``piSort (univSort _9) (λ A → funSort (univSort _7)
+  (funSort _7 (funSort _9 _7)))``
+
+Note that ``funSort`` and ``piSort`` are total functions on sort. But
+``univSort`` is not always well-defined. Eg. without adding the
+``--omega-in-omega`` option, ``Setω`` does not have a ``univSort``
+(successor sort) since there is currently no next sort to ``Setω``.
+Any uses of ``univSort`` will lead to a 'has bigger sort' constraint that
+ensures the argument is not ``Setω``.
+
+.. _set-omega-plus-n:
+
+Sorts ``Setωᵢ``
+---------------
+
+Agda implements sorts of the form ``Setωᵢ``, where ``i`` is a natural number
+``i : Nat``. These sorts constitute a second hierarchy ``Setωᵢ : Setωᵢ₊₁``
```

Comment:
Is really `i : Nat`?  This looks like `i` can be any Agda-expression of type `Nat` (as in `BUILTIN NATURAL Nat`).  Isn't it rather that `i` is a numeral in decimal representation, i.e. `0`, `1`, ... etc.?